pallets/loan/src/lib.rs

```diff
@@ -609,19 +751,86 @@ impl<T: Config> Pallet<T> {
 		let nowt = T::Time::now();
 		TryInto::<u64>::try_into(nowt).map_err(|_| Error::<T>::ErrEpochTimeOverflow.into())
 	}
+
+	/// accrues rate and debt of a given loan and updates it
+	/// returns the present value of the loan
+	fn accrue_and_update_loan(
+		pool_id: T::PoolId,
+		loan_id: T::LoanId,
+		now: u64,
+	) -> Result<T::Amount, DispatchError> {
+		LoanInfo::<T>::try_mutate(
+			pool_id,
+			loan_id,
+			|maybe_loan_data| -> Result<T::Amount, DispatchError> {
+				let mut loan_data = maybe_loan_data.take().ok_or(Error::<T>::ErrMissingLoan)?;
+				let (acc_rate, _debt) = loan_data
+					.accrue(now)
+					.ok_or(Error::<T>::ErrLoanAccrueFailed)?;
+				loan_data.last_updated = now;
+				loan_data.accumulated_rate = acc_rate;
+				let present_value = loan_data
+					.present_value()
+					.ok_or(Error::<T>::ErrLoanPresentValueFailed)?;
+				*maybe_loan_data = Some(loan_data);
+				Ok(present_value)
+			},
+		)
+	}
+
+	/// updates nav for the given pool and returns the latest NAV at this instant
+	fn update_nav(pool_id: T::PoolId) -> Result<T::Amount, DispatchError> {
+		let now = Self::time_now()?;
+		let nav = LoanInfo::<T>::iter_key_prefix(pool_id).try_fold(
+			Zero::zero(),
+			|sum, loan_id| -> Result<T::Amount, DispatchError> {
+				let pv = Self::accrue_and_update_loan(pool_id, loan_id, now)?;
+				sum.checked_add(&pv)
+					.ok_or(Error::<T>::ErrLoanAccrueFailed.into())
+			},
+		)?;
```

Comment:
Is there a bound on the maximal number of loans per pool?

Reply:
no bound at the moment